### PR TITLE
Simplify grammar & syntax (parens) of first class modules

### PR DESCRIPTION
**Pull request by IwanKaramazow**
Fixes https://github.com/facebook/reason/issues/530
Fixes https://github.com/facebook/reason/issues/1715
Fixes https://github.com/facebook/reason/issues/1947
Builds on https://github.com/facebook/reason/pull/1946

### Goals
1. Simplify menhir grammar by over-parsing
2. Allow first-class modules with less parentheses

### Result
* Simpler constraints on let-bindings, uppercase automatically implies a `Ptyp_package`
```reason
/* before */
let thing: (module Thing) = (module MyModule);

/* after */
let thing: Thing = (module MyModule);

/* lowercase stays the same, because it could be a core_type */
let thing: (module Foo.lowercaseModType) = (module MyMod);
```

* Less parentheses
```reason
/* before */
let f = ((module Add : S.Z), x) => Add.add(x);
module Modifier = (val (Db.Hashtbl.create (): (module Db.Sig with type t = Mods.t)));

/* after */
let f = (module Add : S.Z, x) => Add.add(x);
module Modifier = (val Db.Hashtbl.create (): module Db.Sig with type t = Mods.t);
```

* `let module` syntax will need to be deprecated, because of too much conflicts.
```reason
/* before */
let module Lowercase = (val y: x);

/* after */
module Lowercase = (val y: x);
```

Still WIP, checking some extra edge cases and finalizing the code.
Also looking for a way to not make this a breaking change and still support the `let module`-syntax.

### CODE REVIEW

**Comment by chenglou** [comment]
let module is long deprecated and refmting to `module Foo = ...`. So I think it's an ok time to get rid of it

**Comment by IwanKaramazow** [comment]
@chenglou this is ready.
TLDR
* simplify grammar where possible
* less parens for first class modules (parsing + printing)
* `let thing: Thing = (module MyModule);` parses and *prints* as `let thing: module Thing = (module MyModule);`
* no deprecations, `let module` stays as is.

**Comment by chenglou** [comment]
For `let thing: Thing = (module MyModule)`, is it possible to still not accept it? Only `let thing: module Thing = (module MyModule)`?

I'm not too sure that we're at a stage where folks all use refmt yet

**Comment by IwanKaramazow** [comment]
Sure, no problem, I'll restructure the grammar.

File: src/reason-parser/reason_parser.mly [hunk 1]
```diff
@@ -1061,6 +1061,47 @@ let filter_raise_spread_syntax msg nodes =
         raise Reason_syntax_util.(Error(dotdotdotLoc, (Syntax_error msg)))
     | None -> node
     ) nodes
+
+let err loc s =
+  raise Reason_syntax_util.(
```
Comment (chenglou) [inline src/reason-parser/reason_parser.mly]: Do we not have a similar utility already? Genuine question. We should have a single one that throws a single type of exception

File: src/reason-parser/reason_parser.mly [hunk 2]
```diff
@@ -3710,7 +3762,24 @@ mark_position_pat
 pattern_optional_constraint:
 mark_position_pat
   ( pattern                 { $1 }
-  | pattern COLON core_type { mkpat(Ppat_constraint($1, $3)) }
+  | pattern COLON core_type
+    { mkpat(Ppat_constraint($1, $3)) }
+  (* If we kill the `let module …` syntax, this can be placed inside pattern.
```
Comment (chenglou) [inline src/reason-parser/reason_parser.mly]: So apparently @jaredly still uses `let module`. Otherwise I think we can kill `let module`?

**Comment by jaredly** [comment]
I super like `let module` because it's one less difference to memorize between statement-level syntax and expression-level syntax.
If we could make `module X = Y` work at the expression level, I'd be fine with removing `let module` everywhere

**Comment by hcarty** [comment]
@jaredly Where does `module X = Y`  it currently work in expressions? I thought that worked anywhere `let module` worked.

**Comment by jaredly** [comment]
aha I guess it does work :P ok, I'd be ok with it going away.
